src/iter/walk_tree.rs

```diff
@@ -0,0 +1,540 @@
+use crate::iter::plumbing::{bridge_unindexed, Folder, UnindexedConsumer, UnindexedProducer};
+use crate::prelude::*;
+use std::iter::once;
+use std::marker::PhantomData;
+
+#[derive(Debug)]
+struct WalkTreePrefixProducer<'b, S, B, I> {
+    to_explore: Vec<S>, // nodes (and subtrees) we have to process
+    seen: Vec<S>,       // nodes which have already been explored
+    breed: &'b B,       // function generating children
+    phantom: PhantomData<I>,
+}
+
+impl<'b, S, B, I> UnindexedProducer for WalkTreePrefixProducer<'b, S, B, I>
+where
+    S: Send,
+    B: Fn(&S) -> I + Send + Sync,
+    I: IntoIterator<Item = S> + Send,
+{
+    type Item = S;
+    fn split(mut self) -> (Self, Option<Self>) {
+        // explore while front is of size one.
+        while self.to_explore.len() == 1 {
+            let front_node = self.to_explore.pop().unwrap();
+            self.to_explore = (self.breed)(&front_node).into_iter().collect();
```

Comment:
this has nothing to do with the sequential folding. this is the task splitting which requires a new vec

Reply:
You have the existing allocation in `self.to_explore`, now empty since the last item was just popped. Why throw that away for a newly collected `Vec`, instead of using the buffer you have?

The split will happen below, regardless. So the difference is "collect a new `Vec`, drop the old one, and split a new one" (one `free` and two `malloc`s), versus "reuse the old `Vec` and split a new one" (one `malloc`).

Reply:
fair enough, i removed the unneeded allocation